src/runtime_src/tools/xclbinutil/KernelUtilities.cxx

```diff
@@ -126,8 +126,30 @@ void buildXMLKernelEntry(const boost::property_tree::ptree& ptKernel,
 
   // -- Extended-data
   // Blindly add it.  In the future add a JSON schema to validate it
-  const boost::property_tree::ptree& ptExtendedData = ptKernel.get_child("extended-data", ptEmpty);
+  boost::property_tree::ptree ptExtendedData = ptKernel.get_child("extended-data", ptEmpty);
   if (!ptExtendedData.empty()) {
+    // the extended-data in dpu json has "subtype" and "functional" attributes
+    // these two attributes can either have string or numeric value
+    // but xml only taks numeric value
+    // for now, hard-code the conversion
+    auto sFunctional = ptExtendedData.get<std::string>("functional", "");
+    if (sFunctional == "DPU") 
+      sFunctional = "0";
+    else if (sFunctional == "PrePost") 
+      sFunctional = "1";
+    else 
+      ;   // no conversion to sFunctional
+    ptExtendedData.put("functional", sFunctional);
+
+    auto sSubType = ptExtendedData.get<std::string>("subtype", "");
+    if (sSubType == "PS") 
```

Comment:
There is duplicate functionality here.  This code is doing the same thing as SectionIPLayout::getFunctional().  This should not be the case.

Suggestion 
Expose the function SectionIPLayout::getFunctional() via the SectionIPLayout header file and include it in this file.

Note: Ideally, it we be better to expose it xclbin.h, but at this time we cannot do so for it is used amongst multiple clients (e.g., XRT apps, XRT drivers, ZOCL deamons, etc.) and doing so to validate correctness would require a large amount of time.  In addition, there is little ROI.

Reply:
add public APIs in SectionIPLayout.h to be used in KernelUtilities.cxx